EZP-30866: Section is updated after location is moved

SerheyDolgushev opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30866](https://jira.ez.no/browse/EZP-30866)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

**Steps to reproduce**
1. Create new sections: "New" and "Old"
2. Create "New" folder, under root location. Set its section to "New".
3. Create "Move me" folder, under the root location. Set its section to "Old".
4. Move "Move me" folder under "New" location.

**Expected result**: "Move me" folder has the same section as before, which is "Old".

**Actual result**: The section for "Move me" folder is updated to the same value as the parent location ("New") section.

Please note, the section was not updated after location move in legacy.

**TODO**:
- [X] Implement feature / fix a bug.
- [X] Implement tests.
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [X] Ask for Code Review.

adamwojs commented:
If I understand correctly this PR reverts https://github.com/ezsystems/ezpublish-kernel/pull/2042

SerheyDolgushev commented:
Yes. Not sure why mentioned PR was submitted, as in legacy section is not updated when the node is moved (unless there is some custom functionality). I tried to test the case from this PR in the latest legacy, and I get the expected result there.

SerheyDolgushev commented:
Just found this code in legacy https://github.com/ezsystems/ezpublish-legacy/blob/2019.03/kernel/classes/ezcontentobjecttreenodeoperations.php#L81-L93. Investigating why it is ignored on my local.

SerheyDolgushev commented:
Sorry, seems like different logic is implemented in legacy and platform regarding section update after the move:
**eZ Platform**: moved location section is updated, if it is the main location (https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Persistence/Legacy/Content/Location/Handler.php#L353)
**eZ Legacy**: moved location section is updated, if
- it is the main location (https://github.com/ezsystems/ezpublish-legacy/blob/2019.03/kernel/classes/ezcontentobjecttreenodeoperations.php#L81)
- AND moved content section is not the same as new parents (https://github.com/ezsystems/ezpublish-legacy/blob/2019.03/kernel/classes/ezcontentobjecttreenodeoperations.php#L86)
- AND move content section is the same as its old parent content section (https://github.com/ezsystems/ezpublish-legacy/blob/2019.03/kernel/classes/ezcontentobjecttreenodeoperations.php#L91)

The last condition for **eZ Legacy** seems a bit buggy, and it might be the main reason for the different functionality in eZ Legacy and eZ Platform.
Anyways, what is the reason to update section on location move? Shouldn't they be two separate operations?

lserwatka commented:
@SerheyDolgushev we had customer request in the past that section is not updated after move operation, so it was confusing and considered as not expected behaviour.

SerheyDolgushev commented:
https://github.com/ezsystems/ezpublish-legacy/pull/1444 was submitted to fix the logic in legacy.

lserwatka commented:
@SerheyDolgushev thank you, then I suggest to close this one.

SerheyDolgushev commented:
Have been talking to @SylvainGuittard last week about this one. And he had some ideas regarding this PR. I want just to double-check with him this one, before closing it.

lserwatka commented:
@SerheyDolgushev I know, I saw new UI, but this is new story (new JIRA ticket), which requires changes in admin ui as well.

SerheyDolgushev commented:
I see. Closing this one.

andrerom commented:
_So besides of the UI story._

I see from the legacy PR that current legacy behaviour changes section, _but_ unlike in platform now ONLY for object id's that have the same section as root node being moved.

So that seems to be a "bug" or inconsistency, should:
A. Platform be changed to do that as well by default _(Pro: Can't remember anyone complaining about this for 16 years since it was written)_
or
B. Legacy be changed as proposed

gggeek commented:
Iirc the fact that section is update when moving main location, but not when moving a 2ndary location, was documented in legacy training material...
As for the 'best'/'more likely to be expected' behaviour: I have no strong opinion... (between either _not_ changing section or only changing it when moving main location). Maybe an interesting thing to consider is: in case we do change section, would we be doing it recursively for all (primary) children?

andrerom commented:
So by now both legacy and platform _does_ apply section recursively if moving main location, the difference is subtle.

- Legacy only changes section for objects in the subtree which:
    - [Has the same section as the root node being moved, others are untouched]
(https://github.com/ezsystems/ezpublish-legacy/blob/0a605628ed5bc2069e5a38c8adc37fb12bf9d0be/kernel/classes/ezcontentobjecttreenode.php#L2838-L2842)
    - [Are main nodes, non main nodes are untouched](https://github.com/ezsystems/ezpublish-legacy/blob/0a605628ed5bc2069e5a38c8adc37fb12bf9d0be/kernel/classes/ezcontentobjecttreenode.php#L6446-L6447)
- Platform overrides it on all, and thus wipes out any other sections being set in the subtree, it also does not care if sub items are main nodes or not


I think we should align platform with what legacy does at it seems far more fine grained and attempt at being intelligent. 

👍 or 👎  @SerheyDolgushev @gggeek @adamwojs @lserwatka @SylvainGuittard

SerheyDolgushev commented:
I think to mimic the legacy is the better option. IMHO, the best option is to not update section after location move at all.

gggeek commented:
Agree with @SerheyDolgushev. Either not update section at all (and allow devs to do that via a custom workflow), or keep legacy behaviour

andrerom commented:
Ok, so updated proposal then based on that:
- Step 1. **[BUG]**: Change Platform to by default behave like legacy (AKA close legacy PR, and reopen this PR but differently)
- Step 2. **[FEATURE]**: Add option in API and all the way down to opt-out of this, if needed expose advance options in UI to let Editor make the decision when he wants something else then the default

yey 👍 or ney 👎  ?

SylvainGuittard commented:
Thanks André for taking the time to look at this topic. I agree with this approach.

lserwatka commented:
Please note we need to combine both at the same time, we can't ship bug fix as we had already customer report about current behaviour.

andrerom commented:
@lserwatka The bug fix won't be what you find here, it will just be adjustments to the platform logic, stuff that was overlooked when the logic was added in #2042. Details above.
